File: Neetcode-150/TwoPointers/trapping_rain_water.java
public class trapping_rain_water {
    public static int trap(int[] heights) {
        /*
        Brute Force O(n^2)

        int max_area = 0, left_max = 0, right_max = 0;

        for (int i = 0; i < heights.length; i++) {
            left_max = heights[i]; 
            right_max = heights[i];

            for (int j = 0; j < i; j++) {
                left_max = Math.max(left_max, heights[j]);
            }
            for (int k = i + 1; k < heights.length; k++) {
                right_max = Math.max(right_max, heights[k]);
            }

            max_area += Math.min(left_max, right_max) - heights[i];
        }

        return max_area;
        */
        int max_area = 0, left = 0, right = heights.length - 1; // variables for our pointers and total area
        int max_left = heights[left], max_right = heights[right]; // variables to update the max left and max right height values

        while (left < right) { // use two-pointer approach to go through the array
            if (max_left <= max_right) { // if the max left height is less than the max right height we can increment it
                left++;
                max_left = Math.max(max_left, heights[left]); // update the max left height
                max_area += max_left - heights[left]; // update area with max - current height
            }
            else { // opposite of above so when the max right height is less than the max left height we can decrement it
                right--;
                max_right = Math.max(max_right, heights[right]); // update the max right height
                max_area += max_right - heights[right]; // update area with max - current height
            }
        }

        return max_area;
    }
    public static void main(String[] args) {
        int[] heights = {0,2,0,3,1,0,1,3,2,1};

        System.out.println(trap(heights));

        // time complexity of this is O(n) as we go through the array once and the space complexity is O(1)
    }
}

File: Neetcode-150/TwoPointers/TrappingRainWater.py
# Time Complexity: O(n)
# Space Complexity: O(1)
'''
Yep, ignore my incredibly verbose and useless bruteforce soln and replace it with this stupidly 
unintuitive optimal soln. Anyways, the time complexity is O(n) because we're using two pointers unintuitively. 
The space complexity is constant because we're not using any extra data structures. 
'''

def trap(height):

    if not height:
        return 0

    l, r = 0, len(height) - 1
    leftMax, rightMax = height[l], height[r]
    res = 0
    while l < r:
        if leftMax < rightMax:
            l += 1
            leftMax = max(leftMax, height[l])
            res += leftMax - height[l]
        else:
            r -= 1
            rightMax = max(rightMax, height[r])
            res += rightMax - height[r]
    return res

    # left, right = 0, 0
    # maxSum = 0

    # while(left < len(height)):

    #     # Run the right pointer until u reach a higher or equal value
    #     if(height[left] != 0):
            
    #         right = left

    #         # Assuming right doesn't hit the end, once it passes this, it should find the 2nd boundary
    #         right += 1
    #         while(right < len(height)):
    #             if(height[right] == 0):
    #                 right += 1
    #             else:
    #                 break

    #         # If it hits the end and no 2nd boundary is found, no new area can be calculated
    #         if(right >= len(height)):
    #             break
            
    #         # If they're side-by-side
    #         if(right == left + 1):
    #             left = right
    #             continue
                

    #         # Now that you have both boundaries, compare, and find current max sum
    #         currMaxSum = 0
    #         if(height[left] > height[right]):
    #             currMaxSum = height[right] * (right - (left + 1))
    #         else:
    #             currMaxSum = height[left] * (right - (left + 1))

    #         # Now begin the two-pointer part
    #         tempLeft = left + 1
    #         tempRight = right - 1
    #         while(tempLeft <= tempRight):

    #             if(tempLeft != tempRight):
    #                 currMaxSum -= (height[tempLeft] + height[tempRight])
    #             else:
    #                 currMaxSum -= height[tempLeft]
    #             tempLeft += 1
    #             tempRight -= 1
            
    #         maxSum += currMaxSum

    #         left = right
    #     else:
    #         left += 1
    #         right += 1

    # return maxSum

print(trap([0,1,0,2,1,0,1,3,2,1,2,1]))
            














